## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

In [62]:
import pandas as pd
credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
print('INFORMATION')
credit_scoring.info()
print('TOP 10 ROWS')
print(credit_scoring.head(10))
#credit_scoring.head()
print('FINAL 10 ROWS')
print(credit_scoring.tail(10))
#credit_scoring.tail()
print('CREDIT SCORING, COUNT OF COLUMN DATA, NON NULL')
credit_scoring.count()

INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
TOP 10 ROWS
   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   


children            21525
days_employed       19351
dob_years           21525
education           21525
education_id        21525
family_status       21525
family_status_id    21525
gender              21525
income_type         21525
debt                21525
total_income        19351
purpose             21525
dtype: int64

### Conclusion

- Created a dataframe "credit_scoring" by reading the file "/datasets/credit_scoring_eng.csv".
- Loaded the file onto excel sheet to check. The file seems to have some erroneous data that needs to be fixed for processing.
- Printed the dataframe info
- Printed top 10 rows
- Printed final 10 rows
- Missing values in days_employed, total_income as shown by .count(). Only 19351 rows have non null data in days_employed, total_income. Total rows are 21525.
- days_employed and total_income is float, can be changed to integer
- Column 'education' has mixed case values, better to change to same case.

07/12/2020
- I tried credit_scoring.head(), and credit_scoring.tail(). This doesn't lead to any printout. I tried again by removing all Print statements. Either head or tail worked. Both don't work. Not sure.

 # Step 2. Data preprocessing

### Processing missing values

In [63]:
#—Identify and fill in missing values.
#credit_scoring.fillna() 

#fill missing values using fillna(), replace negative children count with 0, replace negative days employed with 0
credit_scoring.fillna(0, inplace = True) 
print('TOP 10 ROWS')
print(credit_scoring.head(10))
print('REPLACE NANs with 0s, CREDIT SCORING, COUNT OF COLUMN DATA, NON NULL',credit_scoring.count())

print('DEBT VALUES',credit_scoring['debt'].unique())

TOP 10 ROWS
   children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40

### Conclusion

Replaced missing values with 0.
07/13 - Why replacing missing values with 0:
Fill missing values using fillna(). Missing values are in two columns (days_employed, total_income) as per the dataframe rows- 
children            21525
days_employed       19351
dob_years           21525
education           21525
education_id        21525
family_status       21525
family_status_id    21525
gender              21525
income_type         21525
debt                21525
total_income        19351
purpose             21525

These two columns are numeric, and have null values. Hence replacing them with 0(zero) for further data analysis.

After filling missing values, all column data is non null as reflected by .count(). All columns have data in 21525 rows.
Printed possible values of "Debt" column.
Will watch for negative children and negative days_employed. If they impact the analysis, then further decision will be made.

07/13 - removed Credit_scoring_refined dataframe reference, changed to credit_scoring dataframe.

### Data type replacement

In [64]:
#—Replace the real number data type with the integer type.
try:
    credit_scoring['days_employed'] = credit_scoring['days_employed'].astype(int)
except:
    print('Datatype cannot be changed for days_employed')

try:
    credit_scoring['total_income'] = credit_scoring['total_income'].astype(int)
except:
    print('Datatype cannot be changed for total_income')

print('INFORMATION')
print(credit_scoring.info())
print('TOP 10 ROWS')
print(credit_scoring.head(10))

INFORMATION
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB
None
TOP 10 ROWS
   children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  Secondary Education             1   
3         3

### Conclusion

Changed the data in columns - Days_employed and total_income to integer. Used the try except clause to make sure that the datatype was changed.

### Processing duplicates

In [65]:
#—Delete duplicate data.

total_rows_with_dupes = credit_scoring['debt'].count()
print("Rows count with dupes",total_rows_with_dupes)

credit_scoring=credit_scoring.drop_duplicates().reset_index(drop=True)

total_rows_with_mixeddupes = credit_scoring['debt'].count()
print("Rows count with samecase dupes removed",total_rows_with_mixeddupes)

credit_scoring['education']= credit_scoring['education'].str.lower()
credit_scoring['income_type']= credit_scoring['income_type'].str.lower()
credit_scoring['purpose']= credit_scoring['purpose'].str.lower()

credit_scoring=credit_scoring.drop_duplicates().reset_index(drop=True)
total_rows_without_dupes = credit_scoring['debt'].count()
print('Rows count after mixed case dupes removed',total_rows_without_dupes)
print('total dupes removed',total_rows_with_dupes - total_rows_without_dupes)

Rows count with dupes 21525
Rows count with samecase dupes removed 21471
Rows count after mixed case dupes removed 21454
total dupes removed 71


### Conclusion

Removed Duplicates:
- Initially used drop_duplicates with whatever case was available. Rows reduced from 21525 to 21471
- Then changed string to lowercase in three columns that are string objects and removed duplicates using drop_duplicates.
- Now have 21454 rows. 
- Mixed case data should be changed to same case before removal of dupes and data processing. This saves time and effort.
Or Pandas should provide an option to remove duplicates while ignoring the "Case sensitivity".

- The dataframe "credit_scoring" is getting better - No Null values, No duplicates, Same lowercase, Data errors reported on children and days_employed < 0, datatypes changed for days_employed, income.

- We are ready for categorizing the data for further analysis.

07/13 - Removed keep=False. It was a mistake

### Categorizing Data

In [66]:
#—Categorize the data.
print(credit_scoring['purpose'].value_counts())
print(credit_scoring['children'].value_counts())
print(credit_scoring['education'].value_counts())
print(credit_scoring['family_status'].value_counts())
print(credit_scoring['total_income'].value_counts())

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

In [67]:
# Erroneous rows
erroneous_rows_children_0 = len(credit_scoring[credit_scoring['children']<0].index)
erroneous_rows_children_20 = len(credit_scoring[credit_scoring['children']==20].index)

erroneous_rows_days_employed = len(credit_scoring[credit_scoring['days_employed']<0].index)

print('TOTAL ERRONEOUS ROWS FOR DAYS_EMPLOYED, SPECIFIED LESS THAN ZERO: ' ,erroneous_rows_days_employed)
print('TOTAL ERRONEOUS ROWS FOR CHILDREN, SPECIFIED LESS THAN ZERO: ' , erroneous_rows_children_0)
print('TOTAL ERRONEOUS ROWS FOR CHILDREN, SPECIFIED 20 OR MORE: ' , erroneous_rows_children_20)

TOTAL ERRONEOUS ROWS FOR DAYS_EMPLOYED, SPECIFIED LESS THAN ZERO:  15906
TOTAL ERRONEOUS ROWS FOR CHILDREN, SPECIFIED LESS THAN ZERO:  47
TOTAL ERRONEOUS ROWS FOR CHILDREN, SPECIFIED 20 OR MORE:  76


In [68]:
#categorizing the 'family_size' data
def categorize_family_size(a):
    if  a <= 0 :
        return "a. no children"
    else:
        return "b. children"
            
credit_scoring["family_size"] = credit_scoring['children'].apply(categorize_family_size)
print('FAMILY SIZE CATEGORIES REFINED - FOR CREDIT SCORING')
print(credit_scoring['family_size'].value_counts())

FAMILY SIZE CATEGORIES REFINED - FOR CREDIT SCORING
a. no children    14138
b. children        7316
Name: family_size, dtype: int64


In [69]:
#categorizing the 'income' column data 'NO - 0', 'LOW <50K', 'MEDIUM 50-100K','HIGH 100-150K VERY HIGH >150K'
def categorize_income(a):
    if  a <= 0 :
        return "a. no income"
    elif (a < 50000):
        return "b. low income"
    elif (a < 100000 ):
        return "c. medium income"
    elif (a < 150000):
        return "d. high income"
    elif (a < 250000):
        return "e. very high"
    else:
        #print(a)
        return 'f. extremely high'
            
credit_scoring["income_category"] = credit_scoring['total_income'].apply(categorize_income)
print('INCOME CATEGORIES REFINED - FOR CREDIT SCORING')
print(credit_scoring['income_category'].value_counts())

INCOME CATEGORIES REFINED - FOR CREDIT SCORING
b. low income        18031
a. no income          2103
c. medium income      1221
d. high income          71
e. very high            22
f. extremely high        6
Name: income_category, dtype: int64


In [70]:
#—Categorize the data.

#Categorizing the 'purpose' column data
def categorize_purpose(a):
    if 'car' in a:
        return "car_purchase"
    elif "wedding" in a:
        return "finance_wedding"
    elif "real" in a or "house" in a or "housing" in a or "property" in a or "rent" in a:
        return "mortgage"
    elif "construction" in a or "build" in a:
        return "construction"
    elif "education" in a or "university" in a or "educate" in a:
        return "education"
    else: 
        #print(a)
        return "other"
            
credit_scoring["purpose_category"] = credit_scoring['purpose'].apply(categorize_purpose)
print('PURPOSE CATEGORIES REFINED - FOR CREDIT SCORING')
print(credit_scoring['purpose_category'].value_counts())

credit_scoring_total_records = len(credit_scoring)

PURPOSE CATEGORIES REFINED - FOR CREDIT SCORING
mortgage           10811
car_purchase        4306
education           4013
finance_wedding     2324
Name: purpose_category, dtype: int64


### Conclusion

ERRONEOUS RECORDS REPORTED FOR CORRECTION:
These might be typos in the entry or database,
This is erroneous data and is reported for future handling at the entry or database level.

Found and reported the number of records with Children < 0. This data should be corrected in the database.
The records with total number of children < 0  is 47. -1 may mean  dash 1 (mistakenly put dash)

Found and reported the number of records with Children > 2. This data should be corrected in the database.
The records with total number of children > 20 is 76. 20 may mean 2 (mistakenly changed to 20).

Found and report the number of records with Days_employed < 0. 
The records with total number of days_employed < 0 is 15906

CATEGORIZATION:
Categorization done for familysize based on number of children
Categorization done for income_category based on amount of income
Categorization done for loan purpose_category based on the purpose.

### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [71]:
credit_scoring_debt = credit_scoring.pivot_table(index=['family_size'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

print(credit_scoring_debt)

debt                0     1  debt_ratio
family_size                            
a. no children  13074  1064    0.075258
b. children      6639   677    0.092537


### Conclusion 

I don't see any correlation between debt and having kids. I had categorized the familysize into:
No Children (children <1)
Children (children > 0)

Ratio of debt/no-debt for each category is .08-.09. 
No conclusion between a debt relation to family with/without children.

- Is there a relation between marital status and repaying a loan on time?

In [72]:
credit_scoring_debt = credit_scoring.pivot_table(index=['family_status'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[credit_scoring_debt['debt_ratio']>0.09]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates

debt                   0    1  debt_ratio
family_status                            
civil partnership   3763  388    0.093471
divorced            1110   85    0.071130
married            11408  931    0.075452
unmarried           2536  274    0.097509
widow / widower      896   63    0.065693
HIGHEST %AGE of DEFAULTS 


debt                  0    1  debt_ratio
family_status                           
civil partnership  3763  388    0.093471
unmarried          2536  274    0.097509

### Conclusion

In the family_status category,
"unmarried" category has the highest number of defaults, followed by "Civil partnership". 
Other categories - divorced, married, widow/er also have high default rates (6-7%)

Married category debt% is 9.75%
Civil Partnership debt% is 9.34%

These are significantly high default rates for the banks. 

- Is there a relation between income level and repaying a loan on time?

In [73]:
credit_scoring_debt = credit_scoring.pivot_table(index=['income_category'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[credit_scoring_debt['debt_ratio']>0.08]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates

debt                   0     1  debt_ratio
income_category                           
a. no income        1933   170    0.080837
b. low income      16552  1479    0.082025
c. medium income    1135    86    0.070434
d. high income        67     4    0.056338
e. very high          21     1    0.045455
f. extremely high      5     1    0.166667
HIGHEST %AGE of DEFAULTS 


debt                   0     1  debt_ratio
income_category                           
a. no income        1933   170    0.080837
b. low income      16552  1479    0.082025
f. extremely high      5     1    0.166667

### Conclusion

I had categorized incomes based on:
<=0, No income,
<50K  - Low income
<100K - Medium income
<150K - High income
>=150K - Extremely high income

The extremely high, low and no income group has a high number of defaults.

Order of defaults (high to low) is:
    extremely high 16.6% (however we have very few of these records so need to be cautious)
    low income 8.2%
    no income 8.1% 

- How do different loan purposes affect on-time repayment of the loan?

In [74]:
credit_scoring_debt = credit_scoring.pivot_table(index=['purpose_category'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[credit_scoring_debt['debt_ratio']>0.09]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates

debt                  0    1  debt_ratio
purpose_category                        
car_purchase       3903  403    0.093590
education          3643  370    0.092200
finance_wedding    2138  186    0.080034
mortgage          10029  782    0.072334
HIGHEST %AGE of DEFAULTS 


debt                 0    1  debt_ratio
purpose_category                       
car_purchase      3903  403     0.09359
education         3643  370     0.09220

### Conclusion

Loan purpose of car_purchase and education has the highest default rate:
car_purchase 9.4% 
education    9.2%

All purpose categories are above 7% so we can assume that all categories are equally defaulting.

EVALUATION FOR EDUCATION CATEGORY

In [75]:
credit_scoring_debt = credit_scoring.pivot_table(index=['education'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[credit_scoring_debt['debt_ratio']>0.09]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates

debt                       0       1  debt_ratio
education                                       
bachelor's degree     4972.0   278.0    0.052952
graduate degree          6.0     NaN         NaN
primary education      251.0    31.0    0.109929
secondary education  13808.0  1364.0    0.089902
some college           676.0    68.0    0.091398
HIGHEST %AGE of DEFAULTS 


debt                   0     1  debt_ratio
education                                 
primary education  251.0  31.0    0.109929
some college       676.0  68.0    0.091398

Education : Debt highest for "Primary Education", "Some college" and Secondary Education.

Bachelors degree - lowest debt defaults
Graduate degree - no defaults

EVALUATION FOR COMBINATION CATEGORIES:
    INCOME, EDUCATION, FAMILY STATUS
    
    FIND HIGHEST DEFAULT COMBINATIONS.
    FIND DEFAULT COMBINATIONS FOR 20 or MORE PEOPLE DEFAULTING.

In [109]:
credit_scoring_debt = credit_scoring.pivot_table(index=['income_category','education','family_status'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

#print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[(credit_scoring_debt['debt_ratio']>0.4)]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates

HIGHEST %AGE of DEFAULTS 


debt                                                 0    1  debt_ratio
income_category   education         family_status                      
f. extremely high bachelor's degree married        1.0  1.0         0.5

Combination of the following are high risk outliers as shown above.

In [108]:
credit_scoring_debt = credit_scoring.pivot_table(index=['income_category','education','family_status'], columns=['debt'], aggfunc='size')

credit_scoring_total_debt =  credit_scoring_debt[1] + credit_scoring_debt[0]
credit_scoring_debt['debt_ratio'] = credit_scoring_debt[1]/credit_scoring_total_debt

#print(credit_scoring_debt)

credit_scoring_final_debt_candidates = credit_scoring_debt[(credit_scoring_debt['debt_ratio']>0.1) & (credit_scoring_debt[1]>20)]

print ("HIGHEST %AGE of DEFAULTS ")
credit_scoring_final_debt_candidates


HIGHEST %AGE of DEFAULTS 


debt                                                        0      1  \
income_category education           family_status                      
a. no income    secondary education civil partnership   258.0   39.0   
b. low income   secondary education unmarried          1361.0  182.0   

debt                                                   debt_ratio  
income_category education           family_status                  
a. no income    secondary education civil partnership    0.131313  
b. low income   secondary education unmarried            0.117952

### Step 4. General conclusion

As you see above, Highest %age (based on ratio of debt to total) is in the following:

low income	and extremely high income
secondary education or primary education
married	    

Combining three Highest Risk categories : 
OUTLIERS: EXTREMELY HIGH SALARY, BACHELORS DEGREE, MARRIED

COMBINATION WITH >30 PEOPE DEFAULTING:
Low (or zero) income, Secondary education, Civil partnership or unmarried

"Is there a relation between having kids and repaying a loan on time?"; NO
"Is there a relation between marital status and repaying a loan on time?"; YES, UNMARRIED and "CIVIL PARTNERSHIP" HAVE HIGHEST LOAN DEFAULTS
"Is there a relation between income level and repaying a loan on time?"; YES, LOW_INCOME_LEVEL, NO INCOME HAVE THE HIGH LOAN DEFAULTS. EXTREMELY HIGH INCOME HAS THE HIGHEST RATE OF DEFAULTS.



### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.

In [81]:
print("Thank you for evaluating my project, Looking forward to your feedback")

Thank you for evaluating my project, Looking forward to your feedback
